language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java

```diff
@@ -256,36 +268,356 @@ public void didOpen(Path filePath, DidOpenTextDocumentParams params) throws Work
+        if (fileEvent.getType() == FileChangeType.Created &&
+                (isBallerinaSourceChange || isBallerinaTomlChange || isCloudTomlChange) &&
+                hasDocumentOrToml(filePath, project)) {
+            // Document might already exists when text/didOpen hits before workspace/didChangeWatchedFiles,
+            // Thus, return silently
+            clientLogger.logTrace(
+                    String.format("Operation '%s' File already exits, {fileUri: '%s' event: '%s'} ignored",
+                                  LSContextOperation.WS_WF_CHANGED.getName(),
+                                  fileEvent.getUri(),
+                                  fileEvent.getType().name()));
+            return;
+        }
+
+        if (isBallerinaSourceChange) {
+            handleWatchedBalSourceChange(filePath, fileEvent, projectPair);
+        } else if (isBallerinaTomlChange) {
+            handleWatchedBallerinaTomlChange(filePath, fileEvent, projectPair);
+        } else if (isCloudTomlChange) {
+            handleWatchedCloudTomlChange(filePath, fileEvent, projectPair);
+        } else if (isDependenciesTomlChange) {
+            handleWatchedDependenciesTomlChange(filePath, fileEvent, projectPair);
+        } else {
+            handleWatchedModuleChange(filePath, fileEvent, projectPair);
+        }
+    }
+
+    private Optional<ProjectPair> projectOfWatchedFileChange(Path filePath, FileEvent fileEvent,
+                                                             boolean isBallerinaSourceChange,
+                                                             boolean isBallerinaTomlChange,
+                                                             boolean isDependenciesTomlChange,
+                                                             boolean isCloudTomlChange,
+                                                             boolean isModuleChange) {
+        if (isBallerinaSourceChange) {
+            if (fileEvent.getType() == FileChangeType.Created) {
+                return projectPair(projectRoot(filePath));
+            } else {
+                // DELETED event
+                // First try as a single-file-project
+                Optional<ProjectPair> optProject = projectPair(filePath);
+                if (optProject.isPresent()) {
+                    return optProject;
+                }
+                // Or Else, try as a build-project
+                Path parent = filePath.getParent();
+                if (ProjectConstants.TEST_DIR_NAME.equals(parent.getFileName().toString())) {
+                    // If inside a tests folder, get parent
+                    parent = parent.getParent();
+                }
+                if (ProjectConstants.MODULES_ROOT.equals(parent.getParent().getFileName().toString())) {
+                    // If inside a module folder, get parent
+                    parent = parent.getParent().getParent();
+                }
+                return projectPair(parent);
```

Comment:
Any specific reason that we can't use `getProjectPair` here?

Reply:
`projectPair` is a direct map lookup. `getProjectPair` can cause creating new project. I'll send a commit to change the name `getProjectPair`

Reply:
+1